spec/services/credit_notes/provider_taxes/report_service_spec.rb

```diff
@@ -0,0 +1,154 @@
+    let(:integration_customer) { create(:anrok_customer, integration:, customer:) }
+    let(:response) { instance_double(Net::HTTPOK) }
+    let(:lago_client) { instance_double(LagoHttpClient::Client) }
+    let(:endpoint) { 'https://api.nango.dev/v1/anrok/finalized_invoices' }
+    let(:body) do
+      path = Rails.root.join('spec/fixtures/integration_aggregator/taxes/invoices/success_response_multiple_fees.json')
+      File.read(path)
+    end
+    let(:integration_collection_mapping) do
+      create(
+        :netsuite_collection_mapping,
+        integration:,
+        mapping_type: :fallback_item,
+        settings: {external_id: '1', external_account_code: '11', external_name: ''}
+      )
+    end
+
+    before do
+      integration_collection_mapping
+      fee_subscription
+      fee_charge
+      integration_customer
+
+      allow(LagoHttpClient::Client).to receive(:new).with(endpoint).and_return(lago_client)
+      allow(lago_client).to receive(:post_with_response).and_return(response)
+      allow(response).to receive(:body).and_return(body)
+    end
+
+    context 'when credit note does not exist' do
+      it 'returns an error' do
+        result = described_class.new(credit_note: nil).call
+
+        aggregate_failures do
+          expect(result).not_to be_success
+          expect(result.error.error_code).to eq('credit_note_not_found')
+        end
+      end
+    end
+
+    context 'when credit note is successfully synced' do
+      it 'returns successful result' do
+        result = report_service.call
+
+        aggregate_failures do
+          expect(result).to be_success
+          expect(result.credit_note.id).to eq(credit_note.id)
+        end
+      end
+
+      it 'discards previous tax errors' do
+        expect { report_service.call }
+          .to change(credit_note.error_details.tax_error, :count).from(1).to(0)
+      end
+    end
+
+    context 'when failed result is returned' do
+      let(:body) do
+        path = Rails.root.join('spec/fixtures/integration_aggregator/taxes/invoices/failure_response.json')
+        File.read(path)
+      end
+
+      it 'returns validation error' do
+        result = report_service.call
+
+        aggregate_failures do
+          expect(result).not_to be_success
+          expect(result.error).to be_a(BaseService::ValidationFailure)
+          expect(LagoHttpClient::Client).to have_received(:new).with(endpoint)
+        end
+      end
+
+      it 'resolves old tax error and creates new one' do
+        old_error_id = credit_note.reload.error_details.order(created_at: :asc).last.id
+
+        report_service.call
+
+        aggregate_failures do
+          expect(credit_note.error_details.tax_error.order(created_at: :asc).last.id).not_to eql(old_error_id)
+          expect(credit_note.error_details.tax_error.count).to be(1)
+          expect(credit_note.error_details.tax_error.order(created_at: :asc).last.discarded?).to be(false)
```

Comment:
I think this is more idiomatic

```suggestion
          expect(credit_note.error_details.tax_error.order(created_at: :asc).last).not_to be_discarded
```

However I wondering if it won't be better to just set the expectation the `ErrorDetails::CreateService` is called with the given parameters as you are testing here the implementation of this service (in the whole test scenario). What do you think?

Reply:
updated